File: kotlin/src/main/kotlin/ai/muonium/muongit/Tag.kt
package ai.muonium.muongit

/** A parsed git annotated tag object */
data class Tag(
    val oid: OID,
    val targetId: OID,
    val targetType: ObjectType,
    val tagName: String,
    val tagger: Signature?,
    val message: String
)

/** Parse a tag object from its raw data content */
fun parseTag(oid: OID, data: ByteArray): Tag {
    val text = data.decodeToString()

    var targetId: OID? = null
    var targetType: ObjectType? = null
    var tagName: String? = null
    var tagger: Signature? = null

    val blankIdx = text.indexOf("\n\n")
    val headerSection = if (blankIdx >= 0) text.substring(0, blankIdx) else text
    val message = if (blankIdx >= 0) text.substring(blankIdx + 2) else ""

    for (line in headerSection.split("\n")) {
        when {
            line.startsWith("object ") -> targetId = OID(line.removePrefix("object "))
            line.startsWith("type ") -> targetType = parseObjectTypeName(line.removePrefix("type "))
            line.startsWith("tag ") -> tagName = line.removePrefix("tag ")
            line.startsWith("tagger ") -> tagger = parseSignatureLine(line.removePrefix("tagger "))
        }
    }

    return Tag(
        oid = oid,
        targetId = targetId ?: throw MuonGitException.InvalidObject("tag missing object"),
        targetType = targetType ?: throw MuonGitException.InvalidObject("tag missing type"),
        tagName = tagName ?: throw MuonGitException.InvalidObject("tag missing tag name"),
        tagger = tagger,
        message = message
    )
}

/** Serialize a tag to its raw data representation (without the object header) */
fun serializeTag(
    targetId: OID,
    targetType: ObjectType,
    tagName: String,
    tagger: Signature?,
    message: String
): ByteArray {
    val sb = StringBuilder()
    sb.append("object ").append(targetId.hex).append('\n')
    sb.append("type ").append(objectTypeName(targetType)).append('\n')
    sb.append("tag ").append(tagName).append('\n')
    if (tagger != null) {
        sb.append("tagger ").append(formatSignatureLine(tagger)).append('\n')
    }
    sb.append('\n')
    sb.append(message)
    return sb.toString().toByteArray()
}

internal fun objectTypeName(type: ObjectType): String = when (type) {
    ObjectType.COMMIT -> "commit"
    ObjectType.TREE -> "tree"
    ObjectType.BLOB -> "blob"
    ObjectType.TAG -> "tag"
}

internal fun parseObjectTypeName(name: String): ObjectType? = when (name) {
    "commit" -> ObjectType.COMMIT
    "tree" -> ObjectType.TREE
    "blob" -> ObjectType.BLOB
    "tag" -> ObjectType.TAG
    else -> null
}

File: swift/src/Tag.swift
/// MuonGit - Tag object read/write
import Foundation

/// A parsed git annotated tag object
public struct Tag: Sendable {
    public let oid: OID
    public let targetId: OID
    public let targetType: ObjectType
    public let tagName: String
    public let tagger: Signature?
    public let message: String
}

// MARK: - Parsing

/// Parse a tag object from its raw data content
public func parseTag(oid: OID, data: Data) throws -> Tag {
    guard let text = String(data: data, encoding: .utf8) else {
        throw MuonGitError.invalidObject("tag is not valid UTF-8")
    }

    var targetId: OID?
    var targetType: ObjectType?
    var tagName: String?
    var tagger: Signature?

    let parts = text.split(separator: "\n\n", maxSplits: 1)
    let headerSection = String(parts[0])
    let message = parts.count > 1 ? String(parts[1]) : ""

    for line in headerSection.split(separator: "\n", omittingEmptySubsequences: false) {
        let line = String(line)
        if line.hasPrefix("object ") {
            targetId = OID(hex: String(line.dropFirst(7)))
        } else if line.hasPrefix("type ") {
            targetType = parseObjectTypeName(String(line.dropFirst(5)))
        } else if line.hasPrefix("tag ") {
            tagName = String(line.dropFirst(4))
        } else if line.hasPrefix("tagger ") {
            tagger = parseSignature(String(line.dropFirst(7)))
        }
    }

    guard let target = targetId else {
        throw MuonGitError.invalidObject("tag missing object")
    }
    guard let type = targetType else {
        throw MuonGitError.invalidObject("tag missing type")
    }
    guard let name = tagName else {
        throw MuonGitError.invalidObject("tag missing tag name")
    }

    return Tag(oid: oid, targetId: target, targetType: type, tagName: name, tagger: tagger, message: message)
}

// MARK: - Serialization

/// Serialize a tag to its raw data representation (without the object header)
public func serializeTag(
    targetId: OID,
    targetType: ObjectType,
    tagName: String,
    tagger: Signature?,
    message: String
) -> Data {
    var lines: [String] = []
    lines.append("object \(targetId.hex)")
    lines.append("type \(objectTypeName(targetType))")
    lines.append("tag \(tagName)")
    if let tagger = tagger {
        lines.append("tagger \(formatSignature(tagger))")
    }
    let header = lines.joined(separator: "\n")
    let raw = "\(header)\n\n\(message)"
    return Data(raw.utf8)
}

// MARK: - Helpers

// objectTypeName is defined in ODB.swift

func parseObjectTypeName(_ name: String) -> ObjectType? {
    return try? objectTypeFromName(name)
}
